# 02. Pandas

Pandas란, 엑셀시트 같은 데이터 조작 및 분석 도구

---

## 주요 특징
- 넘파이를 기반으로 한다
- 1차원 배열 형태의 데이터인 Series, 2차원 표 구조의 데이터인 DataFrame
- 강력한 전처리
- Panel Data (Pandas라는 이름의 어원)

---

## 기본 세팅

In [1]:
import numpy as numpy
import pandas as pd
print(pd.__version__)

3.0.5


---

## Series와 DataFrame

- **Series**: 1차원. 엑셀의 **열 하나**
- **DataFrame**: 2차원 표. **Series 여러 개를 옆으로 붙인 것**

df["Name"]  -> Series  (열 하나를 꺼냄)  
df[["Name"]] -> DataFrame (열 목록을 꺼냄, 결과가 표)

- pd.Series(데이터, index, name): 데이터를 시리즈로 변환
- pd.DataFrame(데이터, index, columns): 데이터를 데이터프레임으로 변환

In [4]:
data = {
    'Name':['Jackson', 'Emma', 'Noah', 'James'],
    'Weight':[68, 74, 77, 78],
    'Option':[True, True, False, False],
    'Rate':[0.21, 1.1, 0.89, 0.91],
    'Type':['A','A','C','C']
}

| Name | Weight | Option | Rate | Type |
|---|---|---|---|---|
| Jackson | 68 | True | 0.21 | A |
| Emma | 74 | True | 1.10 | A |
| Noah | 77 | False | 0.89 | C |
| James | 78 | False | 0.91 | C |


### 데이터프레임으로 변환
- pd.DataFrame(데이터)

In [9]:
df = pd.DataFrame(data)
df

,Name,Weight,Option,Rate,Type
0,Jackson,68,True,0.21,A
1,Emma,74,True,1.10,A
2,Noah,77,False,0.89,C
3,James,78,False,0.91,C


In [8]:
print("df의 타입: ", type(df))

df의 타입:  <class 'pandas.DataFrame'>


### DataFrame -> Numpy 배열

- 데이터프레임.to_numpy()  

컬럼 이름과 인덱스가 사라지고 값만 남는다

In [11]:
df_converted = df.to_numpy()
df_converted

array([['Jackson', 68, True, 0.21, 'A'],
       ['Emma', 74, True, 1.1, 'A'],
       ['Noah', 77, False, 0.89, 'C'],
       ['James', 78, False, 0.91, 'C']], dtype=object)

In [12]:
print("df_converted.dtype: ", df_converted.dtype)

df_converted.dtype:  object


---

## 데이터 분석

- 데이터 첫 5행  
df.head()


- 데이터 마지막 5행  
df.tail()


- 데이터프레임의 크기  
df.shape


- 열 하나 -> Series  
df["Name"]


- 열 목록 -> DataFrame  
df[["Name"]]  
(안쪽 [ ]는 "뽑을 열 이름들의 리스트"이다)

---

## titanic.csv 데이터 분석

> 데이터: Kaggle Titanic 대회의 train.csv를 titanic.csv로 저장해 사용  
> .gitignore(*.csv)로 제외되어 레포에는 포함되지 않음  
> 승객 891명, 12개 컬럼

### 파일에서 읽기
- read_csv(csv파일)

In [13]:
df = pd.read_csv("titanic.csv")
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


### 데이터 분석 (head(), info(), describe() 등등)

#### 1. head()

In [14]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


- 출력에 NaN이 보인다 = 결측치(값이 비어 있음)
- NaN은 "Not a Number". 숫자 컬럼이든 문자 컬럼이든 빈칸은 NaN으로 표시된다

#### 2. info()

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


### info() 읽는 법

| 항목 | 의미 |
|---|---|
| RangeIndex: 891 entries | 전체 891행 |
| Non-Null Count | 값이 있는 행의 개수 |
| Dtype | 자료형 (object = 문자열) |

**전체 891과 다른 컬럼이 결측치가 있는 컬럼이다.**
- Age: 714 -> 177개 비어 있음
- Cabin: 204 -> 687개 비어 있음 (77%)
- Embarked: 889 -> 2개 비어 있음

#### 3. describe()

In [15]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### describe() 읽는 법

- 숫자 컬럼만 나온다 (Name, Sex 같은 문자 컬럼은 제외)
- 25% / 50% / 75% = **사분위수**. 값을 작은 순으로 줄 세웠을 때 1/4, 2/4, 3/4 지점
  - Age 25% = 20.1 -> 승객의 1/4이 20세 이하
  - 50%는 중앙값(median)

### 여기서 읽어낸 것

1. **count가 컬럼마다 다르다** — Age만 714. 결측치가 여기서도 드러난다
2. **Fare는 평균 32.20, 중앙값 14.45, 최대 512.33**
   → 평균이 중앙값의 2배. 소수의 비싼 티켓이 평균을 끌어올렸다
   → 이런 컬럼은 "평균 요금 32달러"라고 말하면 실제와 동떨어진다
3. **Survived의 평균 0.3838 = 생존율 38.4%**
   0/1 값만 있는 컬럼은 평균이 곧 비율이다

#### 4. describe().T

In [16]:
# 전치
df.describe().T

,count,mean,std,min,25%,50%,75%,max
PassengerId,891.0,446.000000,257.353842,1.00,223.5000,446.0000,668.5,891.0000
Survived,891.0,0.383838,0.486592,0.00,0.0000,0.0000,1.0,1.0000
Pclass,891.0,2.308642,0.836071,1.00,2.0000,3.0000,3.0,3.0000
Age,714.0,29.699118,14.526497,0.42,20.1250,28.0000,38.0,80.0000
SibSp,891.0,0.523008,1.102743,0.00,0.0000,0.0000,1.0,8.0000
Parch,891.0,0.381594,0.806057,0.00,0.0000,0.0000,0.0,6.0000
Fare,891.0,32.204208,49.693429,0.00,7.9104,14.4542,31.0,512.3292


컬럼이 많을 때 가로로 길어져서 화면에 다 안 들어오기 때문에 행과 열을 뒤집는 것

In [18]:
# 특정 컬럼의 데이터들만 뽑아서 보기
df[["Age", "Name"]]

,Age,Name
0,22.0,"Braund, Mr. Owen Harris"
1,38.0,"Cumings, Mrs. John Bradley (Florence Briggs Th..."
2,26.0,"Heikkinen, Miss Laina"
3,35.0,"Futrelle, Mrs. Jacques Heath (Lily May Peel)"
4,35.0,"Allen, Mr. William Henry"
...,...,...
886,27.0,"Montvila, Rev. Juozas"
887,19.0,"Graham, Miss Margaret Edith"
888,NaN,"Johnston, Miss Catherine Helen ""Carrie"""
889,26.0,"Behr, Mr. Karl Howell"
